### DISCUSSION TITLE
FIPS check for correct provider seems incorrect

### DISCUSSION THREAD

**Discussion by parsley72** [openssl/openssl#25183] [Q&A]
I'm following the instructions on https://github.com/openssl/openssl/blob/master/README-FIPS.md. I used OpenSSL 3.0.2 (that's what Ubuntu 22.04 currently uses) and OpenSSL 3.0.0 for the FIPS provider. I skipped the section entitled "Use the OpenSSL FIPS provider for testing" but when I got to "Check that the correct FIPS provider is being used" it gave the wrong answer:
```
$ ./util/wrap.pl -fips apps/openssl list -provider-path providers -provider fips -providers
Providers:
  base
    name: OpenSSL Base Provider
    version: 3.0.0
    status: active
  fips
    name: OpenSSL FIPS Provider
    version: 3.0.0
    status: active
```
I realised I was in the OpenSSL 3.0.0 directory because that's where the previous section "Copy the FIPS provider artifacts..." laves you, so I switched to the 3.0.2 directory and tried again:
```
$ ./util/wrap.pl -fips apps/openssl list -provider-path providers -provider fips -providers
Providers:
  base
    name: OpenSSL Base Provider
    version: 3.0.2
    status: active
  fips
    name: OpenSSL FIPS Provider
    version: 3.0.2
    status: active
```
I thought maybe I'd made a mistake so I went back and did the "Use the OpenSSL FIPS provider for testing" section:
```
$ cp ../openssl-3.0.0/providers/fips.so providers/.
$ cp ../openssl-3.0.0/providers/fipsmodule.cnf providers/.
// Validate the output of the following to make sure we are using the
// OpenSSL 3.0.0 FIPS provider
$ ./util/wrap.pl -fips apps/openssl list -provider-path providers \
-provider fips -providers
Providers:
  base
    name: OpenSSL Base Provider
    version: 3.0.2
    status: active
  fips
    name: OpenSSL FIPS Provider
    version: 3.0.0
    status: active
```
Is it that the `apps/openssl` parameter is just looking at the executable in the code directory rather than what's installed on the machine? I tried this:
```
$ ./util/wrap.pl -fips /usr/local/bin/openssl list -provider-path providers -provider fips -providers
Providers:
  base
    name: OpenSSL Base Provider
    version: 3.0.2
    status: active
  fips
    name: OpenSSL FIPS Provider
    version: 3.0.0
    status: active
```
and that seems to give me what I want, but then if I go to the parent directory and run:
```
$ ./openssl-3.0.0/util/wrap.pl -fips /usr/local/bin/openssl list -provider-path providers -provider fips -providers
Providers:
  base
    name: OpenSSL Base Provider
    version: 3.0.0
    status: active
  fips
    name: OpenSSL FIPS Provider
    version: 3.0.0
    status: active
```
it's providing me the results for the openssl-3.0.0 directory; the path of the `openssl` parameter doesn't seem to do anything.

I can't find any documentation for `wrap.pl` but this seems incorrect.

**Comment by t8m** [comment]
The instructions are correct, if they are followed completely. The wrap.pl script ensures that you are running the openssl  application with the libcrypto from the build and not from the system. The instructions say to copy the fips provider along with the fipsmodule.cnf into a build of the primary openssl build version.

If you want to use the fips module from openssl 3.0.0 with the system installed openssl version, you need to find out where are the directories where the system installed openssl version expects the fips provider and overwrite the existing one (from the system).

You then won't use wrap.pl script and run the openssl binary from the system.

**Comment by parsley72** [comment]
I did `sudo make install` for OpenSSL 3.0.2 and the instructions include `sudo make install_fips` for OpenSSL 3.0.0, I thought this would install to the system? But when I run:
```
$ openssl list -provider-path providers -provider fips -providers
list: unable to load provider fips
Hint: use -provider-path option or OPENSSL_MODULES environment variable.
40971AD8717F0000:error:1C8000D5:Provider routines:SELF_TEST_post:missing config data:providers/fips/self_test.c:289:
40971AD8717F0000:error:1C8000E0:Provider routines:ossl_set_error_state:fips module entering error state:providers/fips/self_test.c:387:
40971AD8717F0000:error:1C8000D8:Provider routines:OSSL_provider_init_int:self test post failure:providers/fips/fipsprov.c:706:
40971AD8717F0000:error:078C0105:common libcrypto routines:provider_init:init fail:../crypto/provider_core.c:913:name=fips
```
